common/lib/client/resource.ts

```diff
@@ -0,0 +1,173 @@
+}
+
+function urlFromPathAndParams(path: string, params: Record<string, any>) {
+	return path + (params ? '?' : '') + paramString(params);
+}
+
+function logResponseHandler(callback: Function, method: HttpMethods, path: string, params: Record<string, string>) {
+	return function(err: Error, body: unknown, headers: Record<string, string>, unpacked: boolean, statusCode: number) {
+		if (err) {
+			Logger.logAction(Logger.LOG_MICRO, 'Resource.' + method + '()', 'Received Error; ' + urlFromPathAndParams(path, params) + '; Error: ' + Utils.inspectError(err));
+		} else {
+			Logger.logAction(Logger.LOG_MICRO, 'Resource.' + method + '()',
+				'Received; ' + urlFromPathAndParams(path, params) + '; Headers: ' + paramString(headers) + '; StatusCode: ' + statusCode + '; Body: ' + (BufferUtils.isBuffer(body) ? body.toString() : body));
+		}
+		if (callback) { callback(err, body, headers, unpacked, statusCode); }
+	}
+}
+
+class Resource {
+	static get(rest: Rest, path: string, origheaders: Record<string, string>, origparams: Record<string, any>, envelope: Utils.Format | null, callback: Function): void {
+		Resource.do(HttpMethods.Get, rest, path, null, origheaders, origparams, envelope, callback);
+	} 
+
+	static delete(rest: Rest, path: string, origheaders: Record<string, string>, origparams: Record<string, any>, envelope: Utils.Format | null, callback: Function): void {
+		Resource.do(HttpMethods.Delete, rest, path, null, origheaders, origparams, envelope, callback);
+	} 
+
+	static post(rest: Rest, path: string, body: unknown, origheaders: Record<string, string>, origparams: Record<string, any>, envelope: Utils.Format | null, callback: Function): void {
+		Resource.do(HttpMethods.Post, rest, path, body, origheaders, origparams, envelope, callback);
+	} 
+
+	static patch(rest: Rest, path: string, body: unknown, origheaders: Record<string, string>, origparams: Record<string, any>, envelope: Utils.Format | null, callback: Function): void {
+		Resource.do(HttpMethods.Patch, rest, path, body, origheaders, origparams, envelope, callback);
+	} 
+
+	static put(rest: Rest, path: string, body: unknown, origheaders: Record<string, string>, origparams: Record<string, any>, envelope: Utils.Format | null, callback: Function): void {
+		Resource.do(HttpMethods.Put, rest, path, body, origheaders, origparams, envelope, callback);
+	} 
+
+	static do(method: HttpMethods, rest: Rest, path: string, body: unknown, origheaders: Record<string, string>, origparams: Record<string, any>, envelope: Utils.Format | null, callback: Function): void {
+		if (Logger.shouldLog(Logger.LOG_MICRO)) {
+			callback = logResponseHandler(callback, method, path, origparams);
+		}
+
+		if(envelope) {
+			callback = (callback && unenvelope(callback, envelope));
+			(origparams = (origparams || {}))['envelope'] = envelope;
+		}
+
+		function doRequest(this: any, headers: Record<string, string>, params: Record<string, any>) {
+			if (Logger.shouldLog(Logger.LOG_MICRO)) {
```

Comment:
And this

Reply:
Same as above.

Reply:
Resolving this conversation on @ikbalkaya's behalf.